### ISSUE TITLE
Unite Seoul 2020 Summary

### ISSUE THREAD

**Issue by Haejoon0313** [Haejoon0313/Unity_3D_Game_OneEnemy#1]
# _Day 1_
<br>

### 콘솔 렌더링 파이프라인

- PC, 모바일, PS4
- PBR 기반 라이트
- compute shader를 이용하여 GPU 소모 감소

### Shadow quality 이슈

- 800미터 멀리까지 구현
- 8 cascaded shadow map + position noise soft shadows
- 4 cascade를 매 프레임마다 렌더링 & 나머지 4개는 4프레임마다 번갈아가며 렌더링
- 가장자리의 픽셀만 soft shadow를 계산하고 나머지는 0과 1 반환

### Multiple AO Technologies

- 로컬에서 bake 후 적용

### Local Light Shadows

- 100개의 라이트는 100개의 real-time shadow를 만들어야함
- static의 경우 bake shadow를 이용

### Volumetric Fog & God Rays

- 주변 장애물에 따라 달라지는 안개
- fog를 이용해서 god rays, 즉 빛줄기를 만들어내는것이 가능
- 하지만 fog 외에 새로운 pass를 만드는 이유는 2가지
	- fog의 해상도가 부족
	- fog의 밀도에 따라 ray의 강도가 결정되기 때문에 의도한 형태를 만들기 어려움

### Reflection Probes

- 반사 데이터
- min g-buffer를 이용하여 큐브맵을 만들어서 구현
- 반사광에 의한 그림자를 만들기 위해 게임 내 시간으로 2시간마다 그림자를 베이크한 이후 2개를 서로 보간

### Indoor & Outdoor

- 2개의 probe를 실내와 실외 버전으로 나누어서 만듬
- interior mask를 적용하여 하드 엣지를 제거하고, 푸른 계열 빛을 제거

<br><br>

### VR/AR 디바이스

- AR과 VR의 공통 기능
  - Display (HMD transform / Rendering)
  - Input (Controller)
- 새로운 장비와 기능을 출시가 빨라 개발 난이도 상승
- Developer Framework 구축을 위해 Prefab을 적극 활용

### XR Interaction Toolkit

- 모바일 제스처 / 컨트롤러 / 핸드 트래킹
- Escape Room Example
- Distance Grab / Attach / Socket 지정

### New Input System

- 기존에는 컨트롤러 디바이스 키 맵핑 Table 형태
- XR Controller를 통해 디바이스 키 맵핑을 만들어냄

### AR Foundation

- 현재 4가지 메인 AR 툴
  - Google: ARCore
  - Apple: ARKit
  - Magic Leap
  - HoloLens
- Meshing(iOS), Depth(iOS, Android)

### Unity MARS

- 에디터에서 개발 진행이 어렵고, 빌드 후 디바이스 테스트를 통해 디버깅 과정
- 실제 세계 데이터를 유니티 에디터에 시뮬레이션화해서 테스트 가능
- WYSIWYG 환경

<br><br>

### UI 변경점

- Graph Inspector의 Node Settings에서 Properties 설정 가능
- Master Node -> Master Stack : 기존 렌더링 파이프라인과 유사하게 구조가 바뀌어서 데이터 흐름을 보기 편해짐
- Lit / Unlit / Custom 등에 따라 Input의 종류가 자동으로 변화
- Custom으로 ShaderGraph에서 변경하더라도 코드에 자동으로 반영되지 않음
- 과거 마스터 노드의 설정을 통해 변경하던 설정값들은 Graph Inspector의 Graph Settings에서 변경
- Graph를 따로 만들 필요 없이 그래프 안에서 Material Type을 수정 가능
- Clear Coat 추가 : Object에 투명 막을 코팅한 것과 같은 효과

### Shader Code 변경점

- DOTS 렌더러에서 라이트맵 정보를 받기 위한 properties들이 hide된 채로 추가
- Graph 내부에서 라이팅모드를 정할 수 있기 떄문에 MaterialType이라는 태그 추가
- 버추얼 텍스쳐를 위한 TextureSTack.hlsl include
- Shader Pass 관리를 위해 ShaderPass.hlsl include
- Lit 쉐이더의 경우 Deferred를 위한 GBuffer Pass와 SSAO를 위한 DepthNormals Pass 추가됨

### 새로 추가된 노드

- Parallax Mapping
  - Heightmap을 기반으로 UVmap에 왜곡 부여
  - 얼음의 깊이감 추가
  - 무겁지 않은 편
- Parallax Occulsion Mapping
  - 입체감과 같은 효과를 추가로 고려
  - 소파의 입체감 추가
  - 많이 무거운 편

### Deprecated Nodes

- Shader Graph 10 버전 이전에 만들어진 Color 노드를 대체해서 사용
- 이전 버전 그대로 작동시키길 원하는 경우: Edit > Preference > ShaderGraph에서 Enable Deprecated Nodes 체크
- Deprecated Color 노드에 Colorspace Conversion 노드를 Linear to RGB로 설정하여 연결하면 ShaderGraph 10의 Color와 동일하게 작동함

<br><br>

### KitBash3D Kits

- 대형 영화 제작사 및 AAA 게임 스튜디오에서 사용 중인 에셋
- 다양한 테마의 에셋 모음
- 주로 Pre-Render 프로젝트에 사용됨
- DCC 툴에서 각 에셋을 별도의 fbx 파일로 추출
- LOD Generation을 활용하여 작업
- 현재 에셋 스토어에서 구매 가능

### Lighting

- HDRP : for high-end real time production
- Visual Effects : Chromatic Aberration, Film Grain, Vignette => 과한 고품질의 이미지로 인한 비현실성을 제거하기 위한 장치
- Reflections : Layer system을 이용하여 원하는 오브젝트만 캡쳐
- Contact Shadow : 짧은 스크린 공간 그림자를 서로 가까이 있는 오브젝트에 적용하는 방법
- Volumetric Fog : 하늘과 지면의 차이를 명확히 드러내는 데에 도움이 되며 Scene의 깊이감 조절
- Local Volume : 특정 영역에만 사용하는 추가 보정 => 너무 많이 사용하면 리소스에 악영향

<br><br>

-----------------------

<br><br>

# _Day 2_
<br>

### Object Pooling
- 동일한 객체를 자주 생성/삭제 시 GC의 동작이 많아져 CPU 사용량 증가
- 사용 객체를 충분한 개수만큼 생성해 두었다 사용이 필요한 순간 꺼내 쓰고 사용 후 반납
- Pool의 사이즈를 넘는 양을 활용할 시 보조 조치 필요

### Scriptable Object
- 게임에서 변하지 않는 데이터를 위한 클래스를 작성하는 경우 MonoBehaviour보다 ScriptableObject를 상속하여 객체 작성을 추천
- 메모리 사용량이 비교적 적고 불필요한 Start와 Update 콜이 감소
- MonoBehaviour와 같이 Inspector 사용가능 (Static 객체처럼 사용 가능)

### Variables / Properties
- 동일한 환경의 경우 Properties 대신 Variables 사용 추천
- Properties도 함수 콜을 사용하기 때문에 스택 프레임에서 오버헤드가 발생

### Resources 폴더
- 해당 폴더에 있는 모든 리소스가 사용 여부에 관계 없이 패키징
- 리소스의 개수는 앱 구동 시간에 영향 => Resources 폴더 내의 모든 리소스를 탐색하여 Lookup 테이블을 작성하기 때문에 시간과 메모리 용량이 증가
- 가능하면 Addressable Assets 패키지를 활용하여 AssetBundle을 활용하는 것을 추천

### 비어있는 Unity 이벤트 함수 삭제
- Start, Update 등의 이벤트 함수는 비어있어도 선언되어 잇는 경우 각 이벤트 컨테이너에 추가되고 발생 시에 매번 호출되기 때문에 오버헤드가 발생
- 스크립트의 개수보다 생성된 컴포넌트의 개수가 영향을 많이 주기 때문에 상한 비용 소모
- define 분기를 통해 최종 release에서는 제외

### Awake와 Start 함수는 최대한 가볍게
- 첫 렌더링 결과물이 스크린에 표시되는 시점이 모든 Object의 Start 호출이 완료된 이후이기 떄문에 초기화의 지연은 피해야 함
- 무거운 초기화 로직은 가능하면 Update 내의 별도의 Finite State Machine 등을 통해서 별도의 초기화 구간을 구성하여 진행하는 것을 권장

### Hierarchy 복잡성 감소
- Root GameObject는 모든 하위 오브젝트의 Transform 데이터를 배열로 들고있음
- Transform 배열의 메모리가 재할당되며 모든 배열 값을 다시 설정되며 GC가 발생
- 탐색기의 폴더 방식으로 오브젝트를 관리하기 위한 트리 구조는 지양

### 참조 변수 캐싱
- GameObject.Find => 모든 GameObject를 순회하기 때문에 상당한 오버헤드 발생 (직접 참고 or 초기화 시점에 검색 후 캐싱하여 사용)
- GetComponent, GetComponentInChildren => 마찬가지로 미리 확인 후 캐싱

### Texture Import Settings
- Generate Mipmaps : 밉맵이 필요하지 않은 텍스쳐의 경우 옵션 해제 시 리소스 용량 24% 감소
- Format : 디바이스 하드웨어에서 지원하는 압축 포맷 선택 시 평균적으로 1/4로 감소 (모바일은 OpenCLES 3 이상인 경우 ETC2 / ASTC 포맷이 표준 규격)
- Compression : 동일한 설정도 플랫폼에 따라 결과가 다르기 떄문에 각 플랫폼 별로 확인 필요
- POT 지정 or Atlas 사용 : 명시적으로 POT 값을 지정하여야 낭비 방지
- 불필요시 알파 채널 삭제 : 용량에는 영향이 없으나 이미지 손실률 감소
- Read/Write Enable 해제 : 활성화 시 비디오 메모리에 쓰기용 버퍼를 추가 할당하여 메모리 사용량이 2배로 증가

### Mesh Import Settings
- Mesh Compression : 메시의 불필요한 Vertex를 제거하는 형태로 압축하여 패키지 용량 감소 (Runtime 메모리에는 영향 x) => 적용 후 shape을 확인하여야 함
- Read/Write Enable 해제 : 활성화 시 비디오 메모리에 쓰기용 버퍼를 추가 할당하여 메모리 사용량이 2배로 증가
- Blendshapes, Normals and Tangents : 불필요한 경우 해제하면 Mesh 데이터의 크기 감소
- Rig : 애니메이션을 사용하지 않는 경우 Animation Type을 None으로 설정

### Audio Import Settings
- 가능하면 Force to Mono
- Compression bitrate : 퀄리티 손상을 체크하면서 가능한 bitrate를 낮추면 오디오파일 크기 감소로 패키지 용량과 메모리 사용량 감소
- 추천 포맷 : iOS => ADPCM(효과음), MP3(배경음) / Android => Vorbis
- Clip 크기 별 Load Type 추천 : 200kb 미만 => Decompress On Load / 200kb 이상 => Compressed In Memory / 1mb 이상 => Streaming (압축 해제를 위한 버퍼가 200kb 할당)
- Mute 상태에서는 재생을 중지 => 추가로 가능한 경우 게임 옵션의 Mute 옵션 활성화 시 Audio Clip을 Unload하는 것이 메모리 사용량을 감소시킴

### Batching
- DrawCall 횟수를 줄이면서 렌더링 비용 감소 => Atlas Texture 사용을 권장
- Dynamic Batching : 900개 이하의 Vertex를 가진 오브젝트를 대상으로 한 프레임 당 한번씩 평가하여 합격 시 한번에 그림 => Batching 결과를 프로파일링 하여 성공률이 저조한 경우 옵션을 비활성화하여 CPU 비용 감소
- Static Batching : 빌드 타임에 영향으로 주며, 런타임에는 영향을 주지 않음 => 만약 런타임에 동적으로 Mesh를 추가하려면 StaticBatchingUtility.CombineAPI를 참고하여 Script를 구현해야 함 => 렌더링 비용은 감소되나 Mesh가 메모리 상에 존재하여 메모리 부하 가능성 존재

### Shadow Casting
- 그림자가 필요하지 않은 오브젝트는 명시적으로 Cast Shadow 옵션 비활성화
- Shadow Rendering은 Uniy에서 제공하는 Frame Debugger를 통해서 그림자 계산 결과를 각 오브젝트마다 확인 가능

### Light용 Culling Mask
- Light의 계산 비용은 매우 높음 => 특히 Spot Light는 훨씬 계산양이 높음
- Light 효과를 받는 대상을 별도의 Layer로 구성하여 Culling Mask를 통해 계산 비용을 최소한으로 지정하는 것을 권장

### 모바일 기기 해상도
- 최근 모바일 기기의 해상도는 매우 높은 사양으로 제품 출시 => 이를 그대로 게임 해상도로 사용하는 경우 사용량이 과다
- Screen.SetResolution 함수를 통해 해상도 조절 가능 => GPU 사용량을 줄일 수 있으며 드로우 버퍼 등의 GPU 메모리 사용량도 감소

### UI
- 실제로 보이지 않는 UI도 매 프레임마다 렌더링 => 명시적으로 Activate를 끄거나 Hide 옵션을 설정해서 렌더링 비용 감소
- Sub-Canvas의 UI를 분리시키자 => 정적인 UI와 동적인 UI를 분리
- 하나의 Canvas에 있는 UI의 Z값 동일하게 설정 => 다른 경우 Batching이 깨지며 따로따로 그림
- 같은 texture와 material 사용 => Atlas 적극 활용을 통해 DrawCall 감소
- 동일한 ClippingRect 사용 => 다른 경우 Batching이 이루어지지 않고 별도의 DrawCall로 분리
- 입력 이벤트 처리가 필요한 경우에만 Graphic Raycaster를 활성화 => 입력 이벤트 발생 시 모든 Raycast 대상 UI요소를 확인하므로 이 비용을 방지하기 위해 옵션 비활성화
- Pause나 메인 Menu를 띄울 때 전체 화면 UI 사용 권장 => 게임화면의 Camera를 비활성화하여 렌더링 비용 감소
- 메인 게임 렌더링을 비활성화하고 Frame Rate를 낮추면 CPU와 GPU에게 휴식을 주어 Throttling 현상 예방

<br><br>

### Unity Profiler

- 2020.1 버전부터 추가
- Standalone Process는 에디터 상의 프로세스를 제외하고 모니터링
- Frame Count 옵션 조절 가능
- GPU 프로파일링도 가능 => 디테일에서는 오류가 있을 수 있음 (Built-in 환경에 특화)
- Overview 항목을 CPU와 GPU의 ms를 확인하여 병목 위치 확인 가능
- using UnityEngine.Profiling을 통해 Script에서 제어 가능 (Thread Profiling도 가능)
- Flow Event : Worker에게 Job Distribution 한 부분을 화살표를 통해 확인 가능

### Memory Profiler

- 2020.1.11 버전 기준
- 2개의 Snapshot을 Diff 버튼을 통해 비교 가능 (Memory Map Diff를 통해 그래픽화 가능)
- 큰 변경점 없음

### Frame Debugger

- Why this draw call can't be batched : 41가지 이유를 기준으로 알려줌
- RenderDoc : 오픈소스 => 난독화가 진행된 안드로이드 빌드의 경우 활용이 제한됨

<br><br>

### Unity ML-Agents 1.0

- 기본적으로 강화학습을 이용한 학습 지원
- 강화학습 연구를 위한 시뮬레이션 환경 제작
- 게임 내의 학습 기반 인공지능 NPC 학습하기
- 인공지능을 이용한 게임 스테이지 생성
- 지도학습을 위한 데이터 생성

### Reinforcement Learning

- 보상을 통해 학습을 수행
- 행동을 결정하는 에이전트를 제어 => 보상을 최대화하는 방향으로 행동을 선택하도록 학습

### 알고리즘 구현 방식

- Agent는 Action / Environment는 State와 Reward를 반환
- Agent: Python / Environment: Unity
- 유니티 환경 제작 -> Unity ML-Agents 적용 및 설정 -> 강화학습을 이용하여 에이전트 학습(Unity ML-Agents에서도 지원) -> 학습 완료 시 에이전트 모델을 다시 유니티에 임베딩 -> 유니티 환경을 빌드하여 학습된 에이전트를 적용

### Side Channel

- Unity와 Python 사이에 학습에 직접적으로 연관이 없는 데이터도 교환 가능

### Gym Wrapper

- OpenAI GYM 환경을 통해 에이전트 학습
- ML-Agents로 제작한 환경을 GYM 환경으로 만들어줌

### PyTorch

- 유니티 공식 지원
- 페이스북에서 개발한 딥러닝 라이브러리로 Tensorflow에 비해 구현이 직관적이고 간결
- Release 7부터 사용 가능

### Random Network Distillation

- Release 8부터 사용 가능
- 호기심 기반 탐험 알고리즘 중 Noisy TV 문제를 해결한 알고리즘
- 호기심 기반 탐험 알고리즘 : 랜덤 행동은 비효율적이므로 다음 상태를 알기 어려운 곳으로 에이전트를 유도하는 방식 (보상을 얻기 어려운 Sparce Reward 환경에서 학습)
- 랜덤한 화면이 지속적으로 재생되는 Noisy TV 앞으로만 이동하는 문제 발생
- Montezuma's Revenge 문제 해결

### Google Colab Tutorial

- Release 4부터 사용 가능
- 클라우드 기반의 Jupyter Notebook 개발 환경(Google Research 개발)

<br><br>

-----------------------

<br><br>

# _Day 3_
<br>

### VR 프로그램 최적화

- 90 fps 이상 필요(11ms) => 작업시간(리소스) + 낮은 퀄리티 + 하드웨어
- PBR Textures 활용
- 포인트 클라우드를 기반으로 모델링
- Triplanar Mapping을 이용한 저비용 쉐이더
- Ambient Occulusion : unwrap된 uv를 이용하여 미리 구워서 텍스쳐로 적용 => 불가능한 경우 Vertex Color를 이용하여 AO 적용
- Texture Tiling : 적은 용량으로 화면을 채우기에 효과적이지만 반복이 눈에 보임 => Inigo Quilez를 이용하여 랜덤하게 texture 적용

<br><br>

### 레벨 디자인 영역에서의 우선 순위
- 유저의 행동을 유도하는 것이 레벨 디자인으로 묶여서 표현
- 하프라이프-알릭스는 레벨 디자인이 매우 우수함
- 물체의 크기나 색상도 이목을 끄는 것에 큰 영향
- 명도에 따라 가장 많은 영향
- The Last of Us 2는 수영을 배우는 과정이 시나리오 형태의 레벨 디자인으로 만듬

<br><br>
